pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -210,49 +248,83 @@ func (t *ResolvedPipelineRunTask) conditionsSkip() bool {
 	return false
 }
 
-func (t *ResolvedPipelineRunTask) whenExpressionsSkip(facts *PipelineRunFacts) bool {
+// skipBecauseWhenExpressionsEvaluatedToFalse confirms that the when expressions have completed evaluating, and
+// it returns true if any of the when expressions evaluate to false
+func (t *ResolvedPipelineRunTask) skipBecauseWhenExpressionsEvaluatedToFalse(facts *PipelineRunFacts) bool {
 	if t.checkParentsDone(facts) {
-		if len(t.PipelineTask.WhenExpressions) > 0 {
-			if !t.PipelineTask.WhenExpressions.HaveVariables() {
-				if !t.PipelineTask.WhenExpressions.AllowsExecution() {
-					return true
-				}
-			}
+		if !t.PipelineTask.WhenExpressions.AllowsExecution() {
+			return true
 		}
 	}
 	return false
 }
 
-func (t *ResolvedPipelineRunTask) parentTasksSkip(facts *PipelineRunFacts) bool {
+// skipBecauseParentTaskWasSkipped loops through the parent tasks and checks if the parent task skipped:
+//    if yes, is it because of when expressions and are when expressions?
+//        if yes, it ignores this parent skip and continue evaluating other parent tasks
+//        if no, it returns true to skip the current task because this parent task was skipped
+//    if no, it continues checking the other parent tasks
+func (t *ResolvedPipelineRunTask) skipBecauseParentTaskWasSkipped(facts *PipelineRunFacts) bool {
 	stateMap := facts.State.ToMap()
 	node := facts.TasksGraph.Nodes[t.PipelineTask.Name]
 	for _, p := range node.Prev {
-		if stateMap[p.Task.HashKey()].Skip(facts) {
+		parentTask := stateMap[p.Task.HashKey()]
+		if parentSkipStatus := parentTask.Skip(facts); parentSkipStatus.IsSkipped {
+			// if the `when` expressions are scoped to task and the parent task was skipped due to its `when` expressions,
+			// then we should ignore that and continue evaluating if we should skip because of other parent tasks
+			if parentSkipStatus.SkippingReason == WhenExpressionsSkip && facts.ScopeWhenExpressionsToTask {
+				continue
+			}
 			return true
 		}
 	}
 	return false
 }
 
+// skipBecauseResultReferencesAreMissing checks if the task references results that cannot be resolved, which is a
+// reason for skipping the task, and applies result references if found
+func (t *ResolvedPipelineRunTask) skipBecauseResultReferencesAreMissing(facts *PipelineRunFacts) bool {
+	if t.checkParentsDone(facts) && t.hasResultReferences() {
```

Comment:
hey @jerop this check needs to be revisited. Its causing a task without any when expressions or its parent without having any when expression to skip in case a result reference is missing. The alpha test failure is legit causing a task to be skipped:

```
               "skippedTasks": [
                    {
                        "name": "task2"
                    }
                ],
```

Task2 should fail here since it has a result reference but the results were not initialized by the produce task.

Reply:
The [alpha](https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/4085/pull-tekton-pipeline-alpha-integration-tests/1425292326049157120/) failure: 

```
Error waiting for PipelineRun to fail: "pipelinerun-with-failing-step" completed successfully, should have been failed with reason "InvalidTaskResultReference"
```

I am putting this PR on hold until we fix this check, hopefully for not too long 🤞 

/hold

Reply:
thanks @pritidesai! added a check that it applies to finally tasks and if parent was skipped due to when expression only -- so won't apply to the onError alpha feature

Reply:
Awesome thanks a bunch @jerop 🙏👍